python/ray/util/sgd/v2/worker_group.py

```diff
@@ -1,32 +1,137 @@
+
+    ``WorkerGroup`` launches Ray actors according to the given
+    specification. It can then execute arbitrary Python functions in each of
+    these workers.
+
+    If not enough resources are available to launch the actors, the Ray
+    cluster will automatically scale up if autoscaling is enabled.
+
+    Args:
+        num_workers (int): The number of workers (Ray actors) to launch.
+            Defaults to 1.
+        num_cpus_per_worker (float): The number of CPUs to reserve for each
+            worker. Fractional values are allowed. Defaults to 1.
+        num_gpus_per_worker (float): The number of GPUs to reserve for each
+            worker. Fractional values are allowed. Defaults to 0.
+
+    Example:
+
+        .. code_block:: python
+
+        worker_group = WorkerGroup(num_workers=2)
+        output = worker_group.execute(lambda: 1)
+        assert len(output) == 2
+        assert all(o == 1 for o in output)
+    """
+
     def __init__(self,
                  num_workers: int = 1,
-                 num_cpus_per_worker: int = 1,
-                 num_gpus_per_worker: int = 0):
+                 num_cpus_per_worker: float = 1,
+                 num_gpus_per_worker: float = 0):
+
+        if num_workers <= 0:
+            raise ValueError("The provided `num_workers` must be greater "
+                             f"than 0. Received num_workers={num_workers} "
+                             f"instead.")
+        if num_cpus_per_worker < 0 or num_gpus_per_worker < 0:
+            raise ValueError("The number of CPUs and GPUs per worker must "
+                             "not be negative. Received "
+                             f"num_cpus_per_worker={num_cpus_per_worker} and "
+                             f"num_gpus_per_worker={num_gpus_per_worker}.")
+
+        self.num_workers = num_workers
+        self.num_cpus_per_worker = num_cpus_per_worker
+        self.num_gpus_per_worker = num_gpus_per_worker
         self.workers = self._start_workers()
 
     def _start_workers(self):
-        pass
+        remote_cls = ray.remote(
+            num_cpus=self.num_cpus_per_worker,
+            num_gpus=self.num_gpus_per_worker)(BaseWorker)
+        return [remote_cls.remote() for _ in range(self.num_workers)]
+
+    def restart(self):
+        """Restarts all the workers in this worker group."""
+        self.workers = self._start_workers()
+
+    def shutdown(self, graceful_shutdown_timeout_s: float = 5):
+        """Shutdown all the workers in this worker group.
+
+        Args:
+            graceful_shutdown_timeout_s (float): Attempt a graceful shutdown
+                of the workers for this many seconds. Fallback to force kill
+                if graceful shutdown is not complete after this time.
+        """
+        done_refs = [w.__ray_terminate__.remote() for w in self.workers]
+        # Wait for actors to die gracefully.
+        done, not_done = ray.wait(
+            done_refs, timeout=graceful_shutdown_timeout_s)
```

Comment:
nit: skip this call entirely if `graceful_shutdown_timeout_s` is 0.

Reply:
done